app/connectors/httpParsers/UpscanInitiateHttpParser.scala

```diff
@@ -0,0 +1,48 @@
+
+package connectors.httpParsers
+
+import models.upscan.UpscanInitiateResponseModel
+import play.api.http.Status.{BAD_REQUEST, OK}
+import play.api.libs.json.JsSuccess
+import uk.gov.hmrc.http.{HttpReads, HttpResponse}
+import utils.Logger.logger
+
+object UpscanInitiateHttpParser {
+  type UpscanInitiateResponse = Either[ErrorResponse, UpscanInitiateResponseModel]
+
+  implicit object UpscanInitiateResponseReads extends HttpReads[UpscanInitiateResponse] {
+
+    def read(method: String, url: String, response: HttpResponse): UpscanInitiateResponse = {
+      response.status match {
+        case OK =>
+          response.json.validate[UpscanInitiateResponseModel](UpscanInitiateResponseModel.formats) match {
+            case JsSuccess(model, _) => Right(model)
+            case _ => Left(InvalidJson)
+          }
+        case BAD_REQUEST =>
+          logger.debug(s"[UpScanInitiateResponseReads][read]: Bad request returned with reason: ${response.body}")
+          Left(BadRequest)
+        case status =>
+          logger.warn(s"[UpScanInitiateResponseReads][read]: Unexpected response, status $status returned")
+          Left(UnexpectedFailure(status, s"Unexpected response, status $status returned"))
+      }
+    }
+  }
+
+  trait ErrorResponse {
+    val status: Int
+    val body: String
+  }
+
+  case object InvalidJson extends ErrorResponse {
+    override val status: Int = BAD_REQUEST
+    override val body: String = "Invalid JSON received"
+  }
+
+  case object BadRequest extends ErrorResponse {
+    override val status: Int = BAD_REQUEST
+    override val body: String = "incorrect Json body sent"
+  }
+
+  case class UnexpectedFailure(override val status: Int, override val body: String) extends ErrorResponse
+}
```

Comment:
I'm not a massive fan of mixing case object / classes in the same file. If you want to keep it in the same file then you could make ErrorResponse a sealed trait, but you could argue that these objects are quite generic and could be used elsewhere?

**TL:DR; could you make ErrorResponse a sealed trait or move it to a different file?**

Reply:
Changed error response to a sealed trait